drivers/video/hm0360.c

```diff
@@ -0,0 +1,930 @@
+{
+	const struct hm0360_config *config = dev->config;
+	struct video_format fmt;
+	struct hm0360_data *drv_data = dev->data;
+	struct video_frmival_enum fie = {.format = &drv_data->fmt, .discrete = *frmival};
+	int ret;
+	uint32_t pll_cfg = 0;
+	uint32_t osc_div = 0;
+	bool highres = false;
+
+	ret = hm0360_get_fmt(dev, &fmt);
+	if (ret < 0) {
+		LOG_ERR("Can not get format!");
+		return ret;
+	}
+
+	if (fmt.width == 640 && fmt.height == 480) {
+		highres = true;
+	}
+
+	video_closest_frmival(dev, &fie);
+	LOG_DBG("Applying frame interval number %u", fie.index);
+
+	switch (fie.index) {
+	case HM0360_60_FPS:
+		osc_div = highres ? 0x00 : 0x01;
+		break;
+	case HM0360_30_FPS:
+		osc_div = highres ? 0x01 : 0x02;
+		break;
+	case HM0360_15_FPS:
+		osc_div = highres ? 0x02 : 0x03;
+		break;
+	case HM0360_10_FPS:
+		osc_div = 0x03;
+		break;
+	default:
+		LOG_ERR("Unsupported frame interval value");
+		return -ENOTSUP;
+	}
+
+	ret = video_read_cci_reg(&config->bus, HM0360_CCI_PLL1_CONFIG, &pll_cfg);
+	if (ret < 0) {
+		LOG_ERR("Could not read PLL1 Config");
+		return ret;
+	}
+
+	ret = video_write_cci_reg(&config->bus, HM0360_CCI_PLL1_CONFIG, (pll_cfg & 0xFC) | osc_div);
+	if (ret < 0) {
+		LOG_ERR("Could not set PLL1 Config");
+		return ret;
+	}
+
+	drv_data->cur_frmrate = frmival->denominator;
+
+	LOG_DBG("FrameRate selected: %d", frmival->denominator);
+	LOG_DBG("HIRES Selected: %d", highres);
+	LOG_DBG("OSC DIV: 0x%x", osc_div);
+
+	return 0;
+}
+
+static int hm0360_get_frmival(const struct device *dev, struct video_frmival *frmival)
+{
+	struct hm0360_data *drv_data = dev->data;
+
+	frmival->denominator = drv_data->cur_frmrate;
+	frmival->numerator = 1;
+
+	return 0;
+}
+
+static int hm0360_init_controls(const struct device *dev)
+{
+	int ret;
+	struct hm0360_data *drv_data = dev->data;
+	struct hm0360_ctrls *ctrls = &drv_data->ctrls;
+
+	ret = video_init_ctrl(&ctrls->hflip, dev, VIDEO_CID_HFLIP,
+			      (struct video_ctrl_range){.min = 0, .max = 1, .step = 1, .def = 0});
+	if (ret) {
+		return ret;
+	}
+
+	return video_init_ctrl(&ctrls->vflip, dev, VIDEO_CID_VFLIP,
+			       (struct video_ctrl_range){.min = 0, .max = 1, .step = 1, .def = 0});
+}
+
+static int hm0360_init(const struct device *dev)
+{
+	const struct hm0360_config *config = dev->config;
+	struct video_frmival frmival;
+	int ret;
+
+	struct video_format fmt = {
+		.pixelformat = fmts[0].pixelformat,
+		.width = fmts[0].width_min,
+		.height = fmts[0].height_min,
+	};
+
+	if (!i2c_is_ready_dt(&config->bus)) {
+		/* I2C device is not ready, return */
+		return -ENODEV;
+	}
+
+#if DT_ANY_INST_HAS_PROP_STATUS_OKAY(pwdn_gpios)
+	/* Power up camera module */
+	if (config->pwdn.port != NULL) {
+		if (!gpio_is_ready_dt(&config->pwdn)) {
+			return -ENODEV;
+		}
+		ret = gpio_pin_configure_dt(&config->pwdn, GPIO_OUTPUT_ACTIVE);
+		if (ret < 0) {
+			LOG_ERR("Could not clear power down pin: %d", ret);
+			return ret;
+		}
+	}
+#endif
```

Comment:
Like you did for `reset_gpios`, you might want to add a runtime per-instance check for the `pwdn_gpios` in case it is hard-wired to "on":

```c
		if (config->reset.pwdn != NULL) {
			...
		}
```